drivers/i2c/i2c_numaker.c

```diff
@@ -0,0 +1,792 @@
+			I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk);
+		}
+		break;
+	case S_REPEAT_START_STOP:  /* Slave Transmit/Receive Repeat Start or Stop */
+		/* Go slave end */
+		data->slave_xfer.slave_addressed = false;
+		slave_callbacks->stop(slave_config);
+		I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk | I2C_CTL0_AA_Msk);
+		break;
+#endif  /* CONFIG_I2C_TARGET */
+
+	case BUS_RELEASED:  /* Bus Released */
+		/* Ignore the interrupt raised by BUS_RELEASED. */
+		break;
+	default:
+		__ASSERT(false, "Uncaught I2C FSM state");
+		m_numaker_i2c_master_xfer_end(dev, true);
+	}
+}
+
+static int i2c_numaker_init(const struct device *dev)
+{
+	const struct i2c_numaker_config *config = dev->config;
+	struct i2c_numaker_data *data = dev->data;
+	int err = 0;
+	struct numaker_scc_subsys scc_subsys;
+
+	/* Validate this module's reset object */
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("reset controller not ready");
+		return -ENODEV;
+	}
+
+	/* Clean mutable context */
+	memset(data, 0x00, sizeof(*data));
+
+	k_sem_init(&data->lock, 1, 1);
+	k_sem_init(&data->master_xfer.xfer_sync, 0, 1);
+
+	SYS_UnlockReg();
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx = config->clk_modidx;
+	scc_subsys.pcc.clk_src = config->clk_src;
+	scc_subsys.pcc.clk_div = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
+	err = clock_control_on(config->clkctrl_dev, (clock_control_subsys_t) &scc_subsys);
+	if (err != 0) {
+		goto cleanup;
+	}
+	/* Equivalent to CLK_SetModuleClock() */
+	err = clock_control_configure(config->clkctrl_dev,
+				      (clock_control_subsys_t) &scc_subsys,
+				      NULL);
+	if (err != 0) {
+		goto cleanup;
+	}
+
+	/* Configure pinmux (NuMaker's SYS MFP) */
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err != 0) {
+		goto cleanup;
+	}
+
+	/* Reset I2C to default state, same as BSP's SYS_ResetModule(id_rst) */
+	reset_line_toggle_dt(&config->reset);
+
+	err = i2c_numaker_configure(dev, I2C_MODE_CONTROLLER | i2c_map_dt_bitrate(config->bitrate));
+	if (err != 0) {
+		goto cleanup;
+	}
+
+	config->irq_config_func(dev);
+
+cleanup:
+
+	SYS_LockReg();
+	return err;
+}
+
+static const struct i2c_driver_api i2c_numaker_driver_api = {
+	.configure = i2c_numaker_configure,
+	.get_config = i2c_numaker_get_config,
+	.transfer = i2c_numaker_transfer,
+#ifdef CONFIG_I2C_TARGET
+	.target_register = i2c_numaker_slave_register,
+	.target_unregister = i2c_numaker_slave_unregister,
+#endif
+	.recover_bus = i2c_numaker_recover_bus,
+};
+
+#define NUMAKER_CLKCTRL_DEV_INIT(inst)                                         \
+	.clkctrl_dev = DEVICE_DT_GET(DT_PARENT(DT_INST_CLOCKS_CTLR(inst))),
+
+#define NUMAKER_PINCTRL_DEFINE(inst)                                           \
+	PINCTRL_DT_INST_DEFINE(inst);
+#define NUMAKER_PINCTRL_INIT(inst)                                             \
+	.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),
```

Comment:
are these still needed? I imagine they used to be in a conditional and defined to noop at some point but now you may as well drop them and use the expansion directly... makes it a bit easier to follow

Reply:
Yes, it's not need anymore.
Change to the expansion directly at commit fea78870cd3.